### DISCUSSION TITLE
$request->ip() doesnt work behind Cloudflare

### DISCUSSION THREAD

**Discussion by EnelKyss** [laravel/framework#50645] [Ideas]
$request->ip() always return invalid IP address when the server is using cloudflare proxy. 
there should be something like this:
```php
if (isset($_SERVER["HTTP_CF_CONNECTING_IP"])) {
        $_SERVER['REMOTE_ADDR'] = $_SERVER["HTTP_CF_CONNECTING_IP"];
        $_SERVER['HTTP_CLIENT_IP'] = $_SERVER["HTTP_CF_CONNECTING_IP"];
}
```

**Comment by ghost** [comment]
Put this on construct for \App\Http\Middleware\TrustProxies

    public function __construct()
    {
        $proxies = \explode(',', '0.0.0.0/0,::/0');

        if (\is_array($proxies) && \count($proxies) > 0) {
            $this->proxies = $proxies;
        }
    }

Or declare them in the class param directly
    protected $proxies = ['0.0.0.0/0', '::/0'];

This will allow all proxies by the way.

**Reply by EnelKyss** [reply]
Is that safe ? I feel like TrustProxies.php is there for a reason.

**Reply by ghost** [reply]
When you are using proxies and you need to get the client ip, you need to instruct laravel to trust your proxies (if you know their ip address).

**Reply by Fludem** [reply]
Cloudflare also has two headers it applies to the requests

CF-Connecting-IP << This is the clients IP
and X-Forwarded-For << This is the clients IP if it wasn't in place before the request reached cloudflare

So you can also use this document that gives you their IP addresses so you can trust only their proxies if you want to tighten up security, on top of that you can block traffic that hasn't come via cloudflare:

https://developers.cloudflare.com/fundamentals/concepts/cloudflare-ip-addresses/

**Comment by MuscledGeek** [comment]
I ended up adding this middleware:

```
<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Http\Request;
use Symfony\Component\HttpFoundation\Response;

class TrustCloudflare
{
    /**
     * Handle an incoming request.
     *
     * @param  Closure(Request): (Response)  $next
     */
    public function handle(Request $request, Closure $next): Response
    {
        $cloudflare_ip = $request->header('cf-connecting-ip');

        if($cloudflare_ip)
        {
            $request->server->set('REMOTE_ADDR', $cloudflare_ip);
            $request->setTrustedProxies([$request->ip()], Request::HEADER_X_FORWARDED_FOR);
        }

        return $next($request);
    }
}
```

I'm using Laravel 12. Hope this helps.

**Comment by yudin-s** [comment]
Behind Cloudflare this usually comes down to trusted proxy configuration.

If Laravel does not trust the Cloudflare proxy layer, `$request->ip()` will resolve to a Cloudflare edge IP instead of the real client IP. That then affects rate limiting, audit logs, fraud checks, geo logic, and anything else using the request IP.

I would check:

1. `TrustProxies` is enabled for the proxy/CDN path.
2. The app trusts the correct proxy IP ranges.
3. The forwarded header set matches what Cloudflare sends.
4. You are not trusting `X-Forwarded-For` from arbitrary direct traffic.

The last point matters: if the origin is publicly reachable and you blindly trust forwarded headers, clients can spoof their IP. Ideally the origin only accepts traffic from Cloudflare, or the load balancer/CDN is the only trusted proxy hop.

For SaaS/e-commerce apps I also like logging both values during rollout:

```text
client_ip=...
edge_ip=...
xff=...
```

That makes it obvious whether Laravel is using the visitor IP or the CDN IP before you wire it into limits or security rules.
